### PR TITLE
[envars] honor file "coq_environment.txt"

### PR DESCRIPTION
**Pull request by gares**
The motivation of this PR is to be able to "relocate" a coq that was configure and built with another -prefix.
In this way we can use opam to build coq and then put it into an installer. The installer will then be able to
relocate coq by writing COQLIB=path-of-installation.

### CODE REVIEW

**Comment by gares** [comment]
ping @ppedrot

File: lib/envars.ml [hunk 1]
```diff
@@ -12,7 +12,32 @@ open Util
 
 (** {1 Helper functions} *)
 
-let getenv_else s dft = try Sys.getenv s with Not_found -> dft ()
+(* on Windows env variables are not so common and typically set system wide *)
+let parse_env_line l =
+  try Scanf.sscanf l "%[^=]=%S" (fun name value -> Some(name,value))
+  with _ -> None
+
+let getenv_from_file name =
```
Comment (ppedrot) [inline lib/envars.ml:20]: You need to close the channel opened in this function.

File: lib/envars.ml [hunk 2]
```diff
@@ -12,7 +12,32 @@ open Util
...
+      | _ -> find ()
+    in
+      find ()
+  with
+  | Sys_error s -> raise Not_found
+  | End_of_file -> raise Not_found
+
+let system_getenv name =
+  if Sys.win32 then
+    try Sys.getenv name with Not_found -> getenv_from_file name
```
Comment (ppedrot) [inline lib/envars.ml:36]: Why only for Windows? The advertised motivation is also valid under other OSes?

Comment (gares) [reply lib/envars.ml:36]: I was wondering the same thing. @MSoegtropIMC would this be useful also on OSX?

My main concern is that "./coq_environment.txt"  on Unix is a but like looking for troubles, but `/Applications` on OSX is not really Unix...

Comment (gares) [reply lib/envars.ml]: anyway, done

**Comment by gares** [comment]
I did forget +x on the .sh file, rerunning bench

**Comment by gares** [comment]
dear @coq/doc-maintainers if I search the manual for COQLIB I find some hints, but none seems a good place to document this feature. Suggestions?

File: lib/envars.ml [hunk 3]
```diff
@@ -12,7 +12,36 @@ open Util
 
 (** {1 Helper functions} *)
 
-let getenv_else s dft = try Sys.getenv s with Not_found -> dft ()
+let parse_env_line l =
+  try Scanf.sscanf l "%[^=]=%S" (fun name value -> Some(name,value))
+  with _ -> None
+
+let with_ic file f =
+  let ic = open_in file in
```
Comment (herbelin) [inline lib/envars.ml:20]: Could it be that flag `Open_creat` is needed?

On MacOS X this fails in `bash` (bash 4.4) with `coq_environment.txt: No such file or directory`. (However it works with `tcsh`.)

**Comment by ejgallego** [comment]
IMHO the whole endeavor is flawed, I understand that opam switches are just not relocatable as of today, so even if that works for Coq you may find some more problems with other packages.

You should build with the right opam root IMO. It is known that this is a big pain for windows OCaml users, so you may want to ask them for help [for example @nojb ?]

File: lib/envars.ml (hunk 3 above)
Comment (gares) [reply lib/envars.ml:20]: if the file is not there it should fail (and raise Not_found).
Are you saying that I don't catch the right exception?

**Comment by Zimmi48** [comment]
@gares What about documenting this in `INSTALL.md`? At some point this file should probably be merged into the refman, but in the meantime, this is the place to document `configure` related stuff IMHO.

**Comment by gares** [comment]
The win installer will ship Coq, not the other packages, e.g. not `ocamlc`.
I guess I should check coq_makefile reads this also, and coqdep, if they don't already. But that is pretty much it.

**Comment by ejgallego** [comment]
> The win installer will ship Coq, not the other packages, e.g. not `ocamlc`.

You also may ship a lot of plugins, some linking to libs that may not be relocatable, so you will find pain IMO sooner or later.

**Comment by ejgallego** [comment]
For example is elpi relocatable?

**Comment by gares** [comment]
close call, but I believe it is (it uses `Envars.coqlib`, which this patch let you hack):
https://github.com/LPCIC/coq-elpi/blob/a03579e8f3c1d3dfcfe08c3c81a1795b45ede513/src/coq_elpi_vernacular.ml#L370

**Comment by gares** [comment]
I don't think "coq plugins" are not relocatable.
The `elpi` binary is probably not relocatable, but we don't care, since we don't ship it.

**Comment by ejgallego** [comment]
Depends on the set of plugins, but I predict pain. I was actually talking about `elpi`, not Coq elpi.

The thing is I don't understand why don't you tell opam to use the right directory from the beginning, that would save quite a bit of problems?

**Comment by gares** [comment]
> The thing is I don't understand why don't you tell opam to use the right directory from the beginning, that would save quite a bit of problems?

When you compile the platform on windows you can do that.

But if you want to distribute a pre-built coq, then you can't force the user to install it in c:\coq_platform.
This PR is to let the user install it elsewhere, then the installer would write where it was installed in this file.
We could already export COQLIB, but on windows env variables are more "global" that unix, so if you have 2 coq installations switching between then would not be easy.

**Comment by ejgallego** [comment]
> But if you want to distribute a pre-built coq, then you can't force the user to install it in c:\coq_platform.
> This PR is to let the user install it elsewhere, then the installer would write where it was installed in this file.
> We could already export COQLIB, but on windows env variables are more "global" that unix, so if you have 2 coq installations switching between them would not be easy.

Ah I missed the user-chosen path thing sorry. I guess you should suggest to use `c:\coq_platform` and let the user to switch under their own risk. I have heard this doesn't work sadly in general for OCaml , but people is working on making things relocatable; so I'd guess it would worth asking on the OCaml channels.

**Comment by jfehrle** [comment]
By the way, I think serious users may soon come to prefer running Coq in WSL2 (your favorite Linux running in a Microsoft-developed virtual machine) instead of native Windows.  It's very highly compatible with native Linux and builds Coq faster.  opam just works.  However my IDE dies several times a day, which I think is probably due to inadequate XWindows software.  I expect those problems to be resolved within less than a year.

Beginner-level users may prefer the simplicity of native Windows.  WSL2 takes some time to install, disk space, more memory (e.g. I have 32 GB DRAM) and you need to learn a bit about WSL2.

**Comment by gares** [comment]
Yesterday the platform was tested on wsl and worked.

**Comment by gares** [comment]
apparently `$coqc` is ugly, `'somepath'/foo`, and `cp` does not like it. So I tried making it uniform with readlink, but on osx `-f` does not seem to exist. Ideas?

**Comment by herbelin** [comment]
There is for instance a proposal on [Stack overflow](https://stackoverflow.com/questions/1055671/how-can-i-get-the-behavior-of-gnus-readlink-f-on-a-mac) but is it what you are looking for?

**Comment by gares** [comment]
A comment talks about greadlink, I'll try that

**Review by Zimmi48 — APPROVED** [review]

**Comment by gares** [comment]
Good news, I fixed OSX. Now it's the time of dune.

**Comment by gares** [comment]
update: It turns out that COQLIB is already "relocatable": if the path it points to does not exist, then it make one out of the executable path. Still the feature is needed for OCAMLFIND (unsed by coq_makefile).

**Comment by gares** [comment]
Test suite fixed on dune. I'm testing it on window with now.

**Comment by gares** [comment]
OK, it works on window as well. @ppedrot please give a last check and proceed when CI green.

**Review by MSoegtropIMC — APPROVED** [review]
Looks good to me now

File: lib/envars.ml [hunk 4]
```diff
@@ -12,7 +12,32 @@ open Util
...
 
-let getenv_else s dft = try Sys.getenv s with Not_found -> dft ()
+(* on Windows env variables are not so common and typically set system wide *)
+let parse_env_line l =
+  try Scanf.sscanf l "%[^=]=%S" (fun name value -> Some(name,value))
+  with _ -> None
+
+let getenv_from_file name =
+  try
+    let ic = open_in "coq_environment.txt" in
```
[the remark this answers is not in the document]
Comment (gares) [reply lib/envars.ml]: This is an outdated comment I'm afraid, I now compute `base` from the binary path.

**Comment by gares** [comment]
Offline comment by Emilio: the windows installer also needs to craft a conf file for ocamlfind.
I don't recall what the current windows installed does for this.

**Review by ejgallego — COMMENTED** [review]
This PR seems very suspicious:
- if you are using `Sys.executable_name` then this PR is redundant, as noted in the comments
- for ocamlfind, this may be not enough, as indeed you need the right stuff to be in PATH, IMHO, in particular this is only needed by coq_makefile so maybe coq_makefile should read the config for relocation from its own install config?

**Comment by ejgallego** [comment]
> * for ocamlfind, this may be not enough, as indeed you need the right stuff to be in PATH, IMHO, in particular this is only needed by coq_makefile so maybe coq_makefile should read the config for relocation from its own install config?

Moreover, unless you install the OCaml toolchain `ocamlfind` shouldn't be used by coq_makefile, so I am missing something certainly.

**Comment by gares** [comment]
> Moreover, unless you install the OCaml toolchain `ocamlfind` shouldn't be used by coq_makefile, so I am missing something certainly.

Indeed this is true. I don't really need to fix the value of OCAMLFIND right now, since the installer is not shipping ocaml, but while I was at it I fiexd all the "envars" I saw. The `envars` file is a layer on top of `coq_config` to make the values overrideable. For some reasons `ocamlfind` was not, so I fixed it.

**Comment by MSoegtropIMC** [comment]
> Offline comment by Emilio: the windows installer also needs to craft a conf file for ocamlfind.
> I don't recall what the current windows installed does for this.

@gares : the current installer does nothing. In ancient days when the installer did install OCaml, it did something like this, but this is all commented out. The infrastructure for doing this is still there, though. I just need to know which files to write and possibly an example.

**Comment by ejgallego** [comment]
Yeah unfortunately I don't see a way to install an OCaml toolchain which is relocatable as of today, however there is work going on to fix that so hopefully it can be fixed soon.

For example, if you build dune you can today use the right API to be relocatable [thanks to many efforts including Bobot's] , so actually the last piece of setup that still needs to be fixed is the compiler itself I think.

**Comment by ejgallego** [comment]
Well, actually the opam's fdopen windows packages are precompiled and relocatable so they must already have the machinery there.

**Comment by ejgallego** [comment]
> For example, if you build dune you can today use the right API to be relocatable [thanks to many efforts including Bobot's]

The way it works is for programs to use a library that takes cares of these details.

**Comment by MSoegtropIMC** [comment]
@ejgallego : actually OCaml did work reasonably well - I wrote some config file similar to what @gares is doing. The main reason I removed it was that OCaml didn't make a lot of sense without a posix shell back then - I guess this did change with dune.

**Comment by ejgallego** [comment]
So the problem this PR fixes is that `-prefix` does override the already existing `Sys.executable_name` logic. Wouldn't it be simpler then to have it just a fallback?

**Comment by MSoegtropIMC** [comment]
> If so, wouldn't it be simpler then to have it just a fallback?

I think it does make sense to have the following priorities:

- Highest: supplied environment variables (e.g. when users want to use a non standard library on purpose)
- Configuration file in the path of coqc (if it exists)
- configure prefix
- Lowest: heuristics (like using the path of coqc)

The right thing should be pretty high in the priority chain (for something like the platform installer) while heuristics should be at the end of the chain. This approach is the only method I see to have the right thing being done at a proper priority level.

**Comment by ejgallego** [comment]
As for comparison these are the rules that are used by the relocation reference in the OCaml part:
https://github.com/ocaml/dune/pull/1226#issue-213282554

**Comment by gares** [comment]
These seems to talk about ocaml, Will be relevant when we ship ocaml, if we decide to do so. Right now the platform scripts build an opam root from sources where you want, no relocation needed. Then the nsis installer can be used to give a quick mean to provide Coq prebuilt. That installer only ships Coq and need Coq to be reliably relocatable.

**Comment by ejgallego** [comment]
Note that the ocaml part does include a way to locate the `share` and `lib` directories of arbitrary opam packages, so indeed Coq can follow these rules and indeed it does when built with the standard lib layout.

**Comment by gares** [comment]
These are the most relevant rules I could find in the post you link:

> - if we see the variable OPAM_SWITCH_PREFIX, assume we are in an opam2 environment and that libraries are in $OPAM_SWITCH_PREFIX/lib
> - assume that libraries are in $(ocamlc -where)/../lib

One uses an env variable, the other the ocaml compiler. We ditched the former and can't use the latter, since we don't have ocamlc.

All the other rules in the post use ocamlfind or opam, and we have neither.
What are you talking about, exactly?

**Comment by ejgallego** [comment]
> What are you talking about, exactly?

I was giving an example of a similar problem to see if it could be of help for us, maybe not now, but in future releases. We could in fact use `ocamlfind` as IMHO the platform will have to ship it as it is the only reliable way to have robust plugin loading as of today [due to double-linking stuff]

**Comment by gares** [comment]
Today the windows installer does not provide any service based on ocamlfind, so I cant follow this advice for 8.13, hence I will release with this patch in.

As a future plan, I agree loading coq plugins via ocamlfind, so that common dependencies can be loaded once, is the way to go.
At that point in time, Coq will depend on ocamlfind and the windows installed will have to ship it, configure it, otherwise Coq wont work properly.

But this is not going to happen today, not tomorrow IMO. In the meanwhile, this PR provides a trivial way for binary installers to make sure Coq finds its pieces, wherever they install them. For master, I would suggest to merge this patch, so that the next RM does not have to do the same work if the findlib plan is not finalized in 6 months.

**Comment by ppedrot** [comment]
I am unassigning and leaving the decision to merge into master to the next RM.

**Comment by ejgallego** [comment]
Is there anything blocking this? Is a bit hacky but no a big deal in my book, so if nobody opposes I'll merge.

**Comment by ppedrot** [comment]
Since there is no more opposition, let's merge right away.

**Pull request merged**